src/lex/cpp.rs

```diff
@@ -1832,4 +1857,51 @@ h",
         assert!(cpp("#define f(x) #y\nf(0)").any(|x| x.is_err()));
         assert!(cpp("#define f(x) #+\nf(0)").any(|x| x.is_err()));
     }
+
+    #[test]
+    fn builtins_line() {
+        assert_same("__LINE__", "1");
+        assert_same("\n\n\n\n\n\n\n\n\n__LINE__", "10");
+        assert_same(
+            "#ifdef __LINE__
+            1
+            #endif",
+            "1",
+        );
+        assert_same(
+            "
+            
+            #if __LINE__ == 3
+            1
+            #endif",
+            "1",
+        );
+        assert_same(
+            "#define LINE __LINE__
+            
+
+            LINE",
+            "4",
+        );
+    }
+    #[test]
+    fn builtins_file() {
+        let filename = "helloworld.c";
+        let mut cpp = PreProcessorBuilder::new("__FILE__")
+            .filename(filename)
+            .build();
+        let token = cpp.next_non_whitespace().unwrap().unwrap().data;
+        if let Token::Literal(LiteralToken::Str(rcstrs)) = token {
+            assert_eq!(
+                rcstrs.first().unwrap().as_str(),
+                format!("\"{}\"", filename)
+            );
+        } else {
+            panic!();
+        }
+    }
+    #[test]
+    fn builtins_date_time() {
+        todo!();
```

Comment:
I have no clue how to test `__DATE__` and `__TIME__` since they depend on current time.

Reply:
You could create a `OffsetDateTime` the same way you do in the code and make sure the difference between that and the macro one is less than a few seconds.